# Аналитический отчёт о кино на основе данных MovieLens и IMDb


---

### Введение
В отчёте проведён анализ данных из коллекции MovieLens, дополненной информацией из IMDb. Цель — выявить основные тенденции релизов фильмов, распределение жанров, а также определить лидеров по популярности и качеству в каталоге фильмов.

In [ ]:
import pandas as pd
pd.set_option('display.max_colwidth', None)
from movielens import Links, Ratings, Movies, Tags

In [ ]:
obj_m = Movies('movies.csv')
obj_r = Ratings('ratings.csv')
obj_t = Tags('tags.csv')

#### Распределение кинофильмов по году их производства

In [ ]:
(pd.DataFrame({'Год':obj_m.dist_by_release().keys(), 'Кол-во фильмов':obj_m.dist_by_release().values()})
.sort_values('Год', ascending=True).plot(kind='bar', x='Год', y='Кол-во фильмов', figsize=(18, 6)))

Мы заметили, что количество фильмов, выпускамых в год, чётко коррелирует с ускорением темпа нашей жизни


**Таким образом можно выделить несколько периодов:**

- Заря кинематографа 1984 - 1899 г.г - в год снималось не более 10 кинофильмов.
- Период становления кино 1900 - 1926 г.г - в год снималось от 10 до 50 кинофильмов.
- Период поступательного развития отрасли 1926 - 1990 г.г - кино, кроме развлекательной, начинает выполнять еще и культурно - идеологическую роль. Количество фильмов плавно растет со скромных 50 до почти 500 фильмов в год. 
- Первый скачок скорости производства 1990 - 2000 г.г - по мере того, за 10 лет количество фильмов почти достигает 1000 штук в год. Именно этому периоду мы обязаны появлены появлением большинства наиболее успешнх и значимых названий кинофильмов, которые до сих пор привлекают зрителей.
- Второй скачок, 2000 - н.в - количество фильмов, выпускаемых в год продолжает расти, но темп роста снизился, хотя в  абсолютных числах, сейчас за год выпускается больше кино, чем за десять лет в 1970-е



---

#### Распределение кинофильмов по жанрам

In [ ]:
(pd.DataFrame({'Жанр':obj_m.dist_by_genres().keys(), 'Кол-во фильмов':obj_m.dist_by_genres().values()})
.sort_values('Кол-во фильмов', ascending=True).plot(kind='barh', x='Жанр', y='Кол-во фильмов', figsize=(18, 6)))

- На основе анализа данного графика можно сделать вывод, что наибольшее количество фильмов снято в жанре **"Драма"**
- Второе место занимает жанр **"Комедия"**
- Третье место практически поровну поделили **"Триллеры"**, **"Боевики"** и **"Романтическое кино"**. При этом первые две категории содержать в себе практически столько же кинофильмов, как все остальные вместе взятые. Впрочем, необходимо помнить, что в последние пару десятилетий уже не снимают моножанровых фильмов, поэтому данный анализ не вполне корректен

#### Процентное соотношение фильмов в жанре Драма по годам

In [ ]:
(pd.DataFrame({'Год':obj_m.genre_percentage_over_years('Drama').keys(), 'Доля фильмов жанра Драма, %':obj_m.genre_percentage_over_years('Drama').values()})
.sort_values('Год', ascending=True).plot(kind='line', x='Год', y='Доля фильмов жанра Драма, %', figsize=(18, 6)))

#### Процентное соотношение фильмов в жанре Комедия по годам

In [ ]:
(pd.DataFrame({'Год':obj_m.genre_percentage_over_years('Comedy').keys(), 'Доля фильмов жанра Комедия, %':obj_m.genre_percentage_over_years('Comedy').values()})
.sort_values('Год', ascending=True).plot(kind='line', x='Год', y='Доля фильмов жанра Комедия, %', figsize=(18, 6)))

### Анализ меток

#### Топ-10 тэгов по количеству слов в тэге

In [ ]:
pd.DataFrame({'Тэг':obj_t.most_words(10).keys(), 'Количество слов в тэге':obj_t.most_words(10).values()})

---

Таким образом, рейтинг тегов по длине помогает понять, как именно аудитория выражает мнения о фильмах — через краткие яркие ярлыки или через более длинные, раскрывающие контекст фразы. Это может быть полезно для анализа предпочтений зрителей и улучшения систем рекомендаций или фильтрации контента

**На основе приведённых данных можно сделать следующие выводы:**

**Тема критики и недовольства** аудитории заметно прослеживается, например, по следующим тегам:
- "the catholic church is the most corrupt organization in history"
- "2 villain nonexistent or not needed for good story"
- "stop using useless characters for filler"
- "It was melodramatic and kind of dumb"
Такая картина может указывать на частые жалобы на сюжетные и тематические аспекты фильмов.

**Признание и награды** также популярны: теги с указанием на номинации и премии «Оскар» показывают интерес пользователей к качеству и признанию фильмов:
- "06 Oscar Nominated Best Movie - Animation"
- "Oscar (Best Effects - Visual Effects)"
- "Oscar (Best Music - Original Score)"
 
Тег **"Everything you want is here"** говорит о том, что некоторые фильмы воспринимаются как полноформатные и удовлетворяющие разнообразные ожидания зрителей.
Частота упоминания темы **"based on a true story"** говорит об интересе к фильмам, основанным на реальных событиях.

---

#### Топ-10 самых длинных тегов

In [ ]:
pd.DataFrame({'Топ-10 самых длинных тегов':obj_t.longest(10)})

---

#### Топ-10 самых длинных тегов с наибольшим количеством слов

In [ ]:
pd.DataFrame({'Топ-10 самых длинных тегов с наибольшим количеством слов':obj_t.most_words_and_longest(10)})

#### Топ-10 тэгов по количеству запросов

In [ ]:
pd.DataFrame({'Тэг':obj_t.most_popular(10).keys(), 'Количество':obj_t.most_popular(10).values()})

- Тег **«Netflix»** сам по себе показывает, что речь идет о контенте платформы, и он служит как обобщающий тег для всей выборки.
- Самый частый тег — **«atmospheric»**, что говорит о высоком спросе на фильмы и сериалы с насыщенной и выразительной атмосферой, которые способны погрузить зрителя в уникальный мир.
- Далее по популярности идут **«superhero»** и **«thought-provoking»** (по 24), что отражает интерес аудитории к динамичным экшенам с супергероями и одновременно к фильмам, вызывающим глубокие размышления.
- Теги **«funny»** (23), **«Disney»** (23) и **«surreal»** (23) близки по частоте, показывая, что юмор, семейный контент и необычный, сюрреалистический стиль ценятся почти одинаково.
- Теги **«religion»** (22), **«sci-fi»** (21) и **«dark comedy»** (21) немного менее популярны, но все равно в топе, что свидетельствует о наличии спроса на религиозные темы, научную фантастику и тёмный юмор.

#### Тэги Netflix

In [ ]:
pd.DataFrame({'Тэги Netflix':obj_t.tags_with('Netflix')})

---

#### Распределение зрительской оценки фильма по годам

In [ ]:
(pd.DataFrame({'Год':obj_r.movies.dist_by_year().keys(), 'Пользовательская оценка фильма':obj_r.movies.dist_by_year().values()})
.sort_values('Год', ascending=True).plot(kind='bar', x='Год', y='Пользовательская оценка фильма', figsize=(18, 6)))

#### Распределение количества оценок

In [ ]:
df = (pd.DataFrame({'Оценка':obj_r.movies.dist_by_rating().keys(), 'Количество':obj_r.movies.dist_by_rating().values()}))
df['Оценка'] = df['Оценка'].astype('str')
df.sort_values('Оценка', ascending=True).plot(kind='barh', x='Оценка', y='Количество', figsize=(18, 6))

#### Топ-10 фильмов по количеству оценок за все время

In [ ]:
(pd.DataFrame({'Фильм':obj_r.movies.top_by_num_of_ratings(10).keys(), 'Количество оценок':obj_r.movies.top_by_num_of_ratings(10).values()}))

В списке представлены культовые и широко известные фильмы, такие как "Forrest Gump", "Shawshank Redemption", "Pulp Fiction" и "Star Wars", что говорит о их большой популярности и высокой вовлечённости аудитории.

Таким образом, можно сказать, что фильмы с большим количеством оценок пользуются устойчивым вниманием и заслуживают статуса классики или культовых произведений в киноиндустрии

#### Средний рейтинг топ-10 фильмов

In [ ]:
(pd.DataFrame({'Фильм':obj_r.movies.top_by_ratings(10, 'average').keys(),
               'Средний рейтинг фильма':obj_r.movies.top_by_ratings(10, 'average').values()}))

#### Корреляция кол-ва зрителей и среднего рейтинга фильма

In [ ]:
(pd.DataFrame({'Кол-во зрителей':obj_r.users.users_by_avg_median_ratings(10, 'average').keys(),
               'Средний рейтинг фильма':obj_r.users.users_by_avg_median_ratings(10, 'average').values()}))

В целом наблюдается тенденция, что фильмы с **большим количеством зрителей** имеют несколько **ниже средний рейтин**г, тогда как фильмы с **меньшей аудиторией** имеют чуть более **высокий средний рейтинг**.

Это может объясняться тем, что фильмы с широкой массовой аудиторией часто ориентированы на более широкий спектр вкусов, что приводит к более смешанным оценкам. В то же время менее популярные фильмы могут привлекать более узкую и заинтересованную аудиторию, которая ставит более высокие оценки.

#### Корреляция кол-ва оценок и кол-ва зрителей

In [ ]:
(pd.DataFrame({'Кол-во оценок':obj_r.users.users_by_ratings_num(10).keys(),
               'Кол-во зрителей':obj_r.users.users_by_ratings_num(10).values()}))

Между количеством зрителей и количеством оценок есть прямая связь: фильмы с большим числом зрителей обычно получают больше оценок, так как больше людей участвуют в оценивании.

Такая вариативность указывает на разный уровень вовлеченности зрителей в процесс оценки фильмов, что важно учитывать при анализе рейтингов и популярности

---

### Анализ кинофильмов с IMDB

Для корректной работы этого класса необходимо скачать по ячейке уже готовый датасет

In [1]:
from links import Links
# !gdown -O ex_links.tsv 1UeuodwLYR5PJuPQluivrYj5N1gASBCiT

In [2]:
linker = Links("links.csv")

Links class initialization...
'ex_links.tsv' initialization...
successfully initialized
Now cleaning...
You can call all methods now


#### Информация о интересующих вас фильмах через IMDB  
Передайте как аргументы два списка! 
Список фильмов которые интересны и список полей которые хотите увидеть  
MovieId_list и Fields_list  
поддреживаемые поля :  
*"title", "director", "release_date", "runtime", "budget", "cumulative_gross", "production_companies", "countries_of_origin", "languages"* 

In [3]:
linker.get_imdb([1, 2, 14, 13],
                ["title", "director", "budget"])

[['14'],
 ['Nixon - Der Untergang eines Präsidenten'],
 ['Oliver Stone'],
 ['$44,000,000 (estimated)'],
 ['13'],
 ['Balto: Ein Hund mit dem Herzen eines Helden'],
 ['Simon Wells'],
 ['No Data'],
 ['2'],
 ['Jumanji'],
 ['Joe Johnston'],
 ['$65,000,000 (estimated)'],
 ['1'],
 ['Toy Story'],
 ['John Lasseter'],
 ['$30,000,000 (estimated)']]

#### Топ-10 режисеров по количеству фильмов

In [4]:
linker.top_directors(10)

{'Woody Allen': 48,
 'Alfred Hitchcock': 37,
 'Steven Spielberg': 32,
 'Clint Eastwood': 31,
 'Martin Scorsese': 29,
 'Ridley Scott': 24,
 'Ron Howard': 24,
 'Steven Soderbergh': 22,
 'Barry Levinson': 21,
 'Spike Lee': 21}

Среди режиссеров с наибольшим количеством фильмов лидирует **Вуди Аллен** с 48 картинами, за ним следуют **Альфред Хичкок** (37) и **Стивен Спилберг** (32). Эти режиссеры характеризуются высокой продуктивностью на протяжении длительного периода времени. 
Далее идут такие мастера, как **Клинт Иствуд** (31), **Мартин Скорсезе** (29) и **Ридли Скотт** (24), которые наряду с количеством фильмов обладают и значительным влиянием на индустрию.

#### Топ-10 фильмов по бюджету

In [5]:
linker.most_expensive(10)

{'Мстители: Война бесконечности': '$321000000',
 'Звёздные войны: Последние джедаи': '$317000000',
 'Пираты Карибского моря: На краю света': '$300000000',
 'Лига справедливости': '$300000000',
 'Хан Соло: Звездные войны. Истории': '$275000000',
 'Возвращение Супермена': '$270000000',
 'Рапунцель: Запутанная история': '$260000000',
 'Человек-паук 3: Враг в отражении': '$258000000',
 'Джон Картер': '$250000000',
 'Тёмный рыцарь: Возрождение легенды': '$250000000'}

Главные инвестиции идут в крупные франшизы и проекты с широкой узнаваемостью. Это такие фильмы, как **«Мстители: Война бесконечности»**, **«Звёздные войны»** и **«Пираты Карибского моря»**.

Высокий бюджет обоснован ожиданиями массового зрительского интереса и большого кассового успеха. Такие фильмы часто базируются на популярных брендах, комиксах, литературе или предыдущих успешных частях.
Большие бюджеты направлены на фильмы с масштабными спецэффектами и визуальными эффектами, что требует серьезных затрат на технологии, съемку и постпродакшн
Важным фактором является участие звездных актеров и известных режиссеров, которые требуют высоких гонораров, что увеличивает общий бюджет.

#### Топ-10 фильмов по кассовым сборам

In [6]:
linker.most_profitable(10)

{'Аватар': '$2686710708',
 'Титаник': '$2064812968',
 'Звёздные войны: Пробуждение Силы': '$1826310218',
 'Мстители: Война бесконечности': '$1731415039',
 'Мир Юрского периода': '$1521537444',
 'Форсаж 7': '$1325342457',
 'Мстители': '$1300538536',
 'Гарри Поттер и Дары смерти: Часть II': '$1217505340',
 'Мстители: Эра Альтрона': '$1155018048',
 'Чёрная Пантера': '$1149926083'}

Лидируют масштабные франшизы и блокбастеры с сильным фанатским базисом — **«Аватар»**, **«Титаник»**, **«Звёздные войны»**, **«Мстители»**. Это говорит о том, что проекты с узнаваемым брендом и продолжением истории привлекают максимально широкую аудиторию.
Ключевыми факторами высокого кассового успеха являются не только качество производства, но и мощная маркетинговая кампания и широкое распределение по кинотеатрам по всему миру. Чем больше экранов и дольше фильм в прокате, тем выше сборы.

#### Топ-10 фильмов по хронометражу

In [7]:
linker.longest(10)

{'Крестный отец: Трилогия 1901-1980': '583 mins',
 'Тесицюй': '551 mins',
 'O.J.: Made in America': '467 mins',
 'Лучшие из молодых': '374 mins',
 'Двадцатый век': '317 mins',
 'Двигаясь вперед, иногда я видел краткие проблески красоты': '288 mins',
 'Когда наступит конец света': '287 mins',
 'Дюна': '137 mins',
 'Человек': '263 mins',
 'The Greatest Story Ever Told': '260 mins'}

Самые длинные фильмы — это как правило эпические картины или документальные проекты, которые рассказывают масштабные, глубокие истории и требуют большого экранного времени. Например, **«Крестный отец: Трилогия»** (583 минуты) и **«O.J.: Made in America»** (467 минут) — это многосерийные повествования или расширенные версии.

#### Топ-10 фильмов по стоимости минуты

In [8]:
linker.top_cost_per_minute(10)

{'Терминатор 2 - 3D': 2857142.86,
 'Рапунцель: Запутанная история': 2600000.0,
 'Лига справедливости': 2500000.0,
 'Мстители: Война бесконечности': 2154362.42,
 'Хороший динозавр': 2150537.63,
 'Люди в чёрном 3': 2122641.51,
 'Звёздные войны: Последние джедаи': 2085526.32,
 'Рождественская история': 2083333.33,
 'В поисках Дори': 2061855.67,
 'Хан Соло: Звездные войны. Истории': 2037037.04}

Таким образом, стоимость минуты фильма тесно связана с жанром и технологическим уровнем производства: чем интенсивнее и сложнее технически фильм, тем выше затраты на каждую минуту экранного времени, что отражает стратегию крупных студий вкладывать средства именно в визуальную и технологическую составляющую проектов.

##### Если вы не скачали наш датасет, создайте свой для работы всех методов.  *(==BONUS==)*
  
Для этого укажите ***movieID*** которые хотите использовать, и ***признаки*** которые хотите, чтобы в нем присутствовали. Их нужно передать в   
Поддерживаемые признаки:  
*"title", "director", "release_date", "runtime", "budget", "cumulative_gross", "production_companies", "countries_of_origin", "languages"*  
  
Так же для работы всех методов, вам нужны все признаки!

In [10]:
linker.create_dataset(
                    [1, 2, 3, 7, 4, 14, 13],
                    ["title", "director", "release_date", "runtime", "budget", "cumulative_gross", "production_companies", "countries_of_origin", "languages", "чепуха"]
                )
# linker.dict_from_data() # Если хотите использовать свой датасет - раскоментируйте
# linker.dataset_path = ... # Если хотите хардово указать путь к датасету вам сюда

Введите название вашего датасета:
14, 13, 7, 4, 3, 2, 1, 
Dataset 'ааа.tsv' created successfull
